src/nnue/layers/affine_transform.h

```diff
@@ -177,6 +186,55 @@ namespace Eval::NNUE::Layers {
         sum = _mm_hadd_epi32(sum, sum);
         output[i] = _mm_cvtsi128_si32(sum);
 
+  #elif defined(USE_SSE2)
+        __m128i sum_lo = _mm_cvtsi32_si128(biases_[i]);
+        __m128i sum_hi = kZeros;
+        const auto row = reinterpret_cast<const __m128i*>(&weights_[offset]);
+        for (IndexType j = 0; j < kNumChunks; ++j) {
+          __m128i row_j = _mm_load_si128(&row[j]);
+          __m128i input_j = _mm_load_si128(&input_vector[j]);
+          __m128i row_signs = _mm_cmpgt_epi8(kZeros, row_j);
+          __m128i extended_row_lo = _mm_unpacklo_epi8(row_j, row_signs);
+          __m128i extended_row_hi = _mm_unpackhi_epi8(row_j, row_signs);
+          __m128i extended_input_lo = _mm_unpacklo_epi8(input_j, kZeros);
+          __m128i extended_input_hi = _mm_unpackhi_epi8(input_j, kZeros);
```

Comment:
I think it's better to unpack on parameters load, as I did in my implementation (independent from yours): https://github.com/sf-x/Stockfish/tree/nnue_no_int8 . But this is for maintainer to decide.

Reply:
Thanks for the review, as I'm not an intrinsics expert, I'd like to see some consensus arise between those that know.

Reply:
That would require an entirely separate code path for 16-bit weights, that's too intrusive for this patch.

Reply:
1) You're forgetting that there are already adds which have to run in parallel with the muls, and the ancient CPU we're discussing might well be unable to do anything more. Assume muls are pipelined, then mul takes a cycle (in the throughput sence) also. So unpacking will add overhead (there's now more unpacking than actual work).
2) A separate code path is IMO only moderately intrusive. See my implementation. Besides, it will allow you to remove all unpacking, so the code complexity might not actually increase much.

Reply:
@Fanael Do you have access to such hardware to measure what's faster?